### PR TITLE
DataSourcePlugins: more generics improvements

### PR DESCRIPTION
**Pull request by ryantxu**
I finally found something that will work with strict typing on.  

In this PR, 
* I added `DSType extends DataSourceApi<TQuery, TOptions>` to many signatures
* I changed `React.ComponentClass` to `React.ComponentType` for some classes.  Explore failed when I changed it so I kept it as ComponentClass

### CODE REVIEW

**Comment by torkelo** [comment]
Can't find where DSType or see why it is needed. Can you explain?

File: packages/grafana-ui/src/types/datasource.ts [hunk 1]
```diff
@@ -164,21 +168,28 @@ export interface DataSourceApi<
...
+  TOptions extends DataSourceJsonData = DataSourceJsonData
+> extends DataSourceApi<TQuery, TOptions> {
   modifyQuery?(query: TQuery, action: QueryFixAction): TQuery;
   getHighlighterExpression?(query: TQuery): string;
   languageProvider?: any;
 }
 
-export interface QueryEditorProps<DSType extends DataSourceApi, TQuery extends DataQuery> {
+export interface QueryEditorProps<
+  DSType extends DataSourceApi<TQuery, TOptions>,
```
Comment (ryantxu) [inline packages/grafana-ui/src/types/datasource.ts]: It is subtle, but this is the key change -- everything else just makes this work:

The DSType before was `DSType extends DataSourceApi` -- but we need it to be `DSType extends DataSourceApi<TQuery, TOptions>`

**Comment by ryantxu** [comment]
It is complicated!  

If you move InputDatasource to the grafana-ui package, you will see type errors.  Or try to define a DataSource in an external package.  The root cause (I think... and it all feels like voodo) is that without the added `DSType`, the functions like `setQueryEditor` would complain that the types do not match because they have properties not defined on the interface.

I was only able to find a solution by putting InputDatasource in the grafana-ui package and banging my head on the wall many many times.

File: public/app/plugins/datasource/input/module.ts [hunk 2]
```diff
@@ -6,6 +6,6 @@ import { InputQueryEditor } from './InputQueryEditor';
 import { InputConfigEditor } from './InputConfigEditor';
 import { InputOptions, InputQuery } from './types';
 
-export const plugin = new DataSourcePlugin<InputQuery, InputOptions>(InputDatasource)
+export const plugin = new DataSourcePlugin<InputDatasource, InputQuery, InputOptions>(InputDatasource)
```
Comment (ryantxu) [inline public/app/plugins/datasource/input/module.ts]: It feels like we should be able to simplify with some flavor of generics capture... but I could not find a solution.  The fact that we are passing in a type as the generic and the instance is a little weird.

**Comment by torkelo** [comment]
I can't come up with a better solution.

What do you think @dprokop?

**Comment by dprokop** [comment]
@torkelo We discussed that briefly with Ryan yesterday and, honestly, nothing else came to my mind when I looked at it later:( But taking how complicated the typings are in the consideration I would see looking into options to simplify these at some point.

**Pull request merged**